mono/sgen/sgen-gc.c

```diff
@@ -2999,13 +3023,187 @@ parse_double_in_interval (const char *env_var, const char *opt_name, const char
 	return TRUE;
 }
 
+static SgenMinor
+parse_sgen_minor (const char *opt)
+{
+	if (!opt)
+		return SGEN_MINOR_DEFAULT;
+
+	if (!strcmp (opt, "simple")) {
+		return SGEN_MINOR_SIMPLE;
+	} else if (!strcmp (opt, "simple-par")) {
+		return SGEN_MINOR_SIMPLE_PARALLEL;
+	} else if (!strcmp (opt, "split")) {
+		return SGEN_MINOR_SPLIT;
+	} else {
+		sgen_env_var_error (MONO_GC_PARAMS_NAME, "Using default instead.", "Unknown minor collector `%s'.", opt);
+		return SGEN_MINOR_DEFAULT;
+	}
+}
+
+static SgenMajor
+parse_sgen_major (const char *opt)
+{
+	if (!opt)
+		return SGEN_MAJOR_DEFAULT;
+
+	if (!strcmp (opt, "marksweep")) {
+		return SGEN_MAJOR_SERIAL;
+	} else if (!strcmp (opt, "marksweep-conc")) {
+		return SGEN_MAJOR_CONCURRENT;
+	} else if (!strcmp (opt, "marksweep-conc-par")) {
+		return SGEN_MAJOR_CONCURRENT_PARALLEL;
+	} else {
+		sgen_env_var_error (MONO_GC_PARAMS_NAME, "Using default instead.", "Unknown major collector `%s'.", opt);
+		return SGEN_MAJOR_DEFAULT;
+	}
+
+}
+
+static SgenMode
+parse_sgen_mode (const char *opt)
+{
+	if (!opt)
+		return SGEN_MODE_DEFAULT;
+
+	if (!strcmp (opt, "balanced")) {
+		return SGEN_MODE_BALANCED;
+	} else if (!strcmp (opt, "throughput")) {
+		return SGEN_MODE_THROUGHPUT;
+	} else if (!strcmp (opt, "pause")) {
+		return SGEN_MODE_PAUSE;
+	} else if (!strcmp (opt, "memory")) {
+		return SGEN_MODE_MEMORY;
+	} else if (!strcmp (opt, "power")) {
+		return SGEN_MODE_POWER;
+	} else {
+		sgen_env_var_error (MONO_GC_PARAMS_NAME, "Using default instead.", "Unknown mode `%s'.", opt);
+		return SGEN_MODE_DEFAULT;
+	}
+}
+
+static void
+init_sgen_minor (SgenMinor minor)
+{
+	switch (minor) {
+	case SGEN_MINOR_SIMPLE:
+		sgen_simple_nursery_init (&sgen_minor_collector, FALSE);
+		break;
+	case SGEN_MINOR_SIMPLE_PARALLEL:
+		sgen_simple_nursery_init (&sgen_minor_collector, TRUE);
+		break;
+	case SGEN_MINOR_SPLIT:
+		sgen_split_nursery_init (&sgen_minor_collector);
+		break;
+	default:
+		g_assert_not_reached ();
+	}
+}
+
+static void
+init_sgen_major (SgenMajor major)
+{
+	switch (major) {
+	case SGEN_MAJOR_SERIAL:
+		sgen_marksweep_init (&major_collector);
+		break;
+	case SGEN_MAJOR_CONCURRENT:
+		sgen_marksweep_conc_init (&major_collector);
+		break;
+	case SGEN_MAJOR_CONCURRENT_PARALLEL:
+		sgen_marksweep_conc_par_init (&major_collector);
+		break;
+	default:
+		g_assert_not_reached ();
+	}
+}
+
+/*
+ * Explicit gc params always have priority. If minor and/or major are set
+ * they will override the collector for the specific mode configuration,
+ * while the other gc params are parsed after we do the initialisations here,
+ * potentially overriding them.
+ */
+static void
+init_sgen_mode (SgenMode mode, SgenMinor minor, SgenMajor major)
```

Comment:
Hey Vlad,

There are a few issues with this commit:

1) Lots of magic numbers (plus on previous commit). All those numbers must go to sgen-conf.h as defines. Things like the default mode, default value for sgen_max_pause_time and sgen_max_pause_margin.

2) Using a GC mode should be exclusionary with manual tweaking. Letting them be mixed makes little sense.

3) I don't think we should add all those modes, specially for those we don't have a clear answer. Power and Memory are examples of superfluous modes we can avoid for now.

4) Some modes, like pause, should be parametric, the use should say what's the goal. 30ms pause time is an arbitrary number you picked that doesn't necessarily reflect what the user can expect to meet.

5) This is missing an update to the man page.

6) I don't agree on the configuration picks.

We can ignore HAVE_CONC_GC_AS_DEFAULT when using gc modes as there's no legacy behind it.

Balanced, I agree on this policy. Its goal is to keep pause times in check at a minimal cost in throughput.

Throughput, this one looks wrong, there's no point in using a concurrent GC when the goal is shorter wall clock. Parallel minor and major should be used. It should be give much larger nursery allowance so it better scales with major size.

Pause, I agree on this policy. Its goal is to try to meet a specific pause time goal at a modest sacrifice in performance.

As we discussed, we'll later have a more aggressive variant of the Pause goal where a bigger performance sacrifice is to be expected.